python/sgl_kernel/moe.py

```diff
@@ -252,3 +217,154 @@ def cutlass_fp4_group_mm(
+    assert w2_scale is None, "current MoE does not support w2_scale"
+    assert a1_scale is None, "current MoE does not support a1_scale"
+    assert a2_scale is None, "current MoE does not support a2_scale"
+    assert block_shape is None, "current MoE does not support block_shape"
+
+    # type check
+    assert hidden_states.dtype == torch.bfloat16, "hidden_states must be bfloat16"
+    assert w1.dtype == torch.bfloat16, "w1 must be bfloat16"
+    assert w2.dtype == torch.bfloat16, "w2 must be bfloat16"
+
+    # Shape check
+    assert hidden_states.ndim == 2, "hidden_states must be 2D"
+    assert (
+        hidden_states.shape[-1] == w1.shape[-1]
+    ), f"hidden_states shape[-1] {hidden_states.shape} must be equal to w1 shape[-2] {w1.shape}"
+    assert (
+        2 * w2.shape[2] == w1.shape[1]
+    ), f"w2 shape[2] {w2.shape[2]} must be half of w1 shape[1] {w1.shape[1]}"
+    assert (topk_ids.shape == topk_weights.shape) and (
+        topk_ids.shape[0] == hidden_states.shape[0]
+    ), f"topk_ids shape {topk_ids.shape} and topk_weights shape {topk_weights.shape} must be equal and match hidden_states shape[0] {hidden_states.shape[0]}"
+
+    num_tokens, _ = hidden_states.shape
+
+    E, _, K = w1.shape
+    E, OutK, N = w2.shape
+    assert N * 2 == w1.shape[1], "w1 shape[1] must be 2x of w2 shape[2]"
+
+    M = num_tokens
+    TopK = topk_ids.shape[1]
+
+    # import pdb; pdb.set_trace()
+    cache = torch.empty(
+        M * TopK * max(2 * N, OutK),
+        device=hidden_states.device,
+        dtype=hidden_states.dtype,
+    )
+    intermediate_cache1 = cache[: M * TopK * 2 * N].view((M * TopK, 2 * N))
+    intermediate_cache2 = torch.empty(
+        (M * TopK, N),
+        device=hidden_states.device,
+        dtype=hidden_states.dtype,
+    )
+    intermediate_cache3 = cache[: M * TopK * OutK].view((M * TopK, OutK))
+
+    if no_combine:
+        assert not inplace
+        out_hidden_states = torch.empty(
+            (num_tokens, OutK),
+            device=hidden_states.device,
+            dtype=hidden_states.dtype,
+        )
+    elif inplace:
+        out_hidden_states = hidden_states
+    else:
+        out_hidden_states = torch.zeros_like(hidden_states)
+
+    flat_topk = topk_ids.flatten()
+    idxs = flat_topk.argsort()
+    sorted_expert_ids = flat_topk[idxs]
+
+    counts = torch.bincount(sorted_expert_ids, minlength=E)  # [E]
+    token_idxs = idxs // TopK  # [num_tokens * TopK]
+    input_A = torch.empty(
+        (num_tokens * TopK, K), device=hidden_states.device, dtype=hidden_states.dtype
+    )
+    input_A = hidden_states[token_idxs].squeeze(1)
+    offset = counts.to(torch.int32)
+
+    torch.ops.sgl_kernel.moe_grouped_mm_nt(intermediate_cache1, input_A, w1, offset, E)
+
+    torch.ops.sgl_kernel.silu_and_mul(intermediate_cache2, intermediate_cache1)
+
+    torch.ops.sgl_kernel.moe_grouped_mm_nt(
+        intermediate_cache3, intermediate_cache2.contiguous(), w2, offset, E
+    )
```

Comment:
there is a trick here to keep the 2nd group gemm using `intermediate_cache2` in the reordered index, so that you only gather `hidden_states` for 1st group gemm.